docs: adds ai apps docs (#1764)
slackapi/python-slack-sdk: docs/english/web.md

See the [`chat.postEphemeral`](/reference/methods/chat.postEphemeral) API method for more details.

### Sending streaming messages {#sending-streaming-messages}

You can have your app's messages stream in to replicate conventional AI chatbot behavior. This is done through three Web API methods:

* [`chat_startStream`](/reference/methods/chat.startstream)
* [`chat_appendStream`](/reference/methods/chat.appendstream)
* [`chat_stopStream`](/reference/methods/chat.stopstream)

> :::tip [The Python Slack SDK provides a [`chat_stream()`](https://docss.slack.dev/tools/python-slack-sdk/reference/web/client.html#slack_sdk.web.client.WebClient.chat_stream) helper utility to streamline calling these methods.]

Reviewer: ```suggestion
:::tip [The Python Slack SDK provides a [`chat_stream()`](/tools/python-slack-sdk/reference/web/client.html#slack_sdk.web.client.WebClient.chat_stream) helper utility to streamline calling these methods.]
```

🔗 suggestion: Can we absolute reference this?

Reviewer 2: fix: There is a double-s in the `docs.` but the relative path makes sense to me!

Author: yeah ill fix the typo - but the full path is needed for python reference docs. relative paths will attempt loading a docusaurus page at that path, which will 404. the python reference docs are technically not pages, but static files that we are hosting using the domain (which are just coincidentally html files that load in a browser)

See the [_Streaming messages_](/tools/bolt-python/concepts/message-sending#streaming-messages) section of the Bolt for Python docs for implementation instructions. 

:::

#### Starting the message stream {#starting-stream}

First you need to begin the message stream:

```python
# Example: Stream a response to any message
@app.message()

> def handle_message(message, client):
>     channel_id = payload["channel"]
>     thread_ts = payload["thread_ts"]
>     
>     # Start a new message stream
>     stream_response = client.chat_startStream(
>         channel=channel_id,
>         thread_ts=thread_ts,
>     )
>     stream_ts = stream_response["ts"]

Reviewer: ```suggestion
def handle_message(message, client):
    channel_id = event.get("channel")
    team_id = event.get("team")
    thread_ts = event.get("thread_ts") or event.get("ts")
    user_id = event.get("user")
    
    # Start a new message stream
    stream_response = client.chat_startStream(
        channel=channel_id,
        recipient_team_id=team_id,
        recipient_user_id=user_id,
        thread_ts=thread_ts,
    )
    stream_ts = stream_response["ts"]
```

🪬 suggestion: Streaming into channel in response to messages outside of DMs will require a specific `recipient` at this time. Also added fallback for threading against top-level messages without a thread timestamp!

Reviewer 2: +1 to this!

```

#### Appending content to the message stream {#appending-stream}

With the stream started, you can then append text to it in chunks to convey a streaming effect.

The structure of the text coming in will depend on your source. The following code snippet uses OpenAI's response structure as an example:

```python
# continued from above
    for event in returned_message:
        if event.type == "response.output_text.delta":
            client.chat_appendStream(
                channel=channel_id, 
                ts=stream_ts, 
                markdown_text=f"{event.delta}"
            )
        else:
            continue
```

#### Stopping the message stream {#stopping-stream}

Your app can then end the stream with the `chat_stopStream` method:

```python
# continued from above
    client.chat_stopStream(
        channel=channel_id, 
        ts=stream_ts
    )
```

> The method also provides you an opportunity to request user feedback on your app's responses using the [feedback buttons](/reference/block-kit/block-elements/feedback-buttons-element) block element within the [context actions](/reference/block-kit/blocks/context-actions-block) block. The user will be presented with thumbs up and thumbs down buttons.

Reviewer: ```suggestion
The method also provides you an opportunity to request user feedback on your app's responses using the [feedback buttons](/reference/block-kit/block-elements/feedback-buttons-element) block element within the [context actions](/reference/block-kit/blocks/context-actions-block) block. The user will be presented with thumbs up and thumbs down buttons which send an action to your app when pressed.
```

💡 question: I forget if this is clear elsewhere, but would it be useful to include a note that these emit actions? I can imagine confusion perhaps in thinking this feedback is saved elsewhere somehow...
